src/main/java/org/springframework/data/r2dbc/repository/query/R2dbcQueryCreator.java

```diff
@@ -0,0 +1,175 @@
+        super(tree);
+        this.tree = tree;
+
+        Assert.notNull(dataAccessStrategy, "Data access strategy must not be null");
+        Assert.notNull(entityMetadata, "Relational entity metadata must not be null");
+        Assert.notNull(parameterMetadataProvider, "Parameter metadata provider must not be null");
+
+        this.dataAccessStrategy = dataAccessStrategy;
+        this.entityMetadata = entityMetadata;
+
+        RenderContext renderContext = dataAccessStrategy.getStatementMapper().getRenderContext();
+        RenderNamingStrategy namingStrategy = renderContext == null
+                ? NamingStrategies.asIs()
+                : renderContext.getNamingStrategy();
+        this.conditionFactory = new ConditionFactory(dataAccessStrategy, namingStrategy, parameterMetadataProvider);
+    }
+
+    /**
+     * Creates {@link Condition} for the given method name part.
+     *
+     * @param part method name part (must not be {@literal null})
+     * @param iterator iterator over query parameter values
+     * @return new instance of {@link Condition}
+     */
+    @Override
+    protected Condition create(Part part, Iterator<Object> iterator) {
+        return conditionFactory.createCondition(part);
+    }
+
+    /**
+     * Combines the given {@link Condition} with the new one created for the given method name part using {@code AND}.
+     *
+     * @param part method name part (must not be {@literal null})
+     * @param base condition to be combined (must not be {@literal null})
+     * @param iterator iterator over query parameter values
+     * @return condition combination
+     */
+    @Override
+    protected Condition and(Part part, Condition base, Iterator<Object> iterator) {
+        return base.and(conditionFactory.createCondition(part));
+    }
+
+    /**
+     * Combines two {@link Condition}s using {@code OR}.
+     *
+     * @param base condition to be combined (must not be {@literal null})
+     * @param condition another condition to be combined (must not be {@literal null})
+     * @return condition combination
+     */
+    @Override
+    protected Condition or(Condition base, Condition condition) {
+        return base.or(condition);
+    }
+
+    /**
+     * Creates SQL query applying the given {@link Condition} and {@link Sort} definition.
+     *
+     * @param condition condition to be applied to query
+     * @param sort      sort option to be applied to query (must not be {@literal null})
+     * @return SQL query
+     */
+    @Override
+    protected String complete(Condition condition, Sort sort) {
+        Table fromTable = Table.create(dataAccessStrategy.toSql(entityMetadata.getTableName()));
+        Collection<? extends Expression> selectExpressions = getSelectionExpressions(fromTable);
+        SelectFromAndJoin selectBuilder = StatementBuilder.select(selectExpressions).from(fromTable);
+
+        if (tree.isExistsProjection()) {
```

Comment:
Nit: `PartTree` exposes a `isLimiting()` method that allows imposing a limit on the actual query (`findTop3`). It would make sense to consider that feature while being at query derivation.